ddsc/core/d4s2.py

```diff
@@ -294,29 +297,123 @@ def _copy_project(self, project_name, new_project_name, path_filter):
+        self._download_project(activity, project_name, temp_directory, path_filter)
+        self._upload_project(activity, new_project_name, temp_directory)
+        activity.finished()
         shutil.rmtree(temp_directory)
         return self.remote_store.fetch_remote_project(new_project_name, must_exist=True)
 
-    def _download_project(self, project_name, temp_directory, path_filter):
+    def _download_project(self, activity, project_name, temp_directory, path_filter):
         """
         Download the project with project_name to temp_directory.
+        :param activity: CopyActivity: info about the copy activity are downloading for
         :param project_name: str name of the pre-existing project
         :param temp_directory: str path to directory we can download into
         :param path_filter: PathFilter: filters what files are shared
         """
         self.print_func("Downloading a copy of '{}'.".format(project_name))
-        downloader = ProjectDownload(self.remote_store, project_name, temp_directory, path_filter)
+        downloader = ProjectDownload(self.remote_store, project_name, temp_directory, path_filter,
+                                     file_download_pre_processor=DownloadedFileRelations(activity))
         downloader.run()
 
-    def _upload_project(self, project_name, temp_directory):
+    def _upload_project(self, activity, project_name, temp_directory):
         """
         Upload the contents of temp_directory into project_name
+        :param activity: CopyActivity: info about the copy activity are uploading for
         :param project_name: str project name we will upload files to
         :param temp_directory: str path to directory who's files we will upload
         """
         self.print_func("Uploading to '{}'.".format(project_name))
         items_to_send = [os.path.join(temp_directory, item) for item in os.listdir(os.path.abspath(temp_directory))]
-        project_upload = ProjectUpload(self.config, project_name, items_to_send)
+
+        project_upload = ProjectUpload(self.config, project_name, items_to_send,
+                                       file_upload_post_processor=UploadedFileRelations(activity))
         project_upload.run()
+
+
+class CopyActivity(object):
+    def __init__(self, data_service, project_name, new_project_name):
+        """
+        Create an activity for our copy operation so users can trace back where the copied files came from.
+        :param data_service: DataServiceApi: service used to create the activity
+        :param project_name: str project name we will download files from
+        :param new_project_name: str project name we will upload files into
+        """
+        self.data_service = data_service
+        self.name = "DukeDSClient copying project: {}".format(project_name)
+        self.desc = "Copying {} to project {}".format(project_name, new_project_name)
```

Comment:
Can we add the version of DukeDSClient to the description, and perhaps the project IDs too (since the names are not unique). May also be a good place to put encode some structured metadata if there are other fields we want to capture. machine name? local user? project size and transfer times? I know the latter two won't be available until completion, just brainstorming.